capsules/extra/src/tickv.rs

```diff
@@ -39,12 +33,165 @@
 use core::cell::Cell;
 use kernel::hil::flash::{self, Flash};
 use kernel::hil::hasher::{self, Hasher};
-use kernel::hil::kv_system::{self, KVSystem};
 use kernel::utilities::cells::{MapCell, OptionalCell, TakeCell};
 use kernel::utilities::leasable_buffer::{SubSlice, SubSliceMut};
 use kernel::ErrorCode;
 use tickv::{self, AsyncTicKV};
 
+/// The type of keys, this should define the output size of the digest
+/// operations.
+pub trait KeyType: Eq + Copy + Clone + Sized + AsRef<[u8]> + AsMut<[u8]> {}
+
+impl KeyType for [u8; 8] {}
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait KVSystemClient<K: KeyType> {
+    /// This callback is called when the append_key operation completes.
+    ///
+    /// - `result`: Nothing on success, 'ErrorCode' on error
+    /// - `unhashed_key`: The unhashed_key buffer
+    /// - `key_buf`: The key_buf buffer
+    fn generate_key_complete(
+        &self,
+        result: Result<(), ErrorCode>,
+        unhashed_key: SubSliceMut<'static, u8>,
+        key_buf: &'static mut K,
+    );
+
+    /// This callback is called when the append_key operation completes.
+    ///
+    /// - `result`: Nothing on success, 'ErrorCode' on error
+    /// - `key`: The key buffer
+    /// - `value`: The value buffer
+    fn append_key_complete(
+        &self,
+        result: Result<(), ErrorCode>,
+        key: &'static mut K,
+        value: SubSliceMut<'static, u8>,
+    );
+
+    /// This callback is called when the get_value operation completes.
+    ///
+    /// - `result`: Nothing on success, 'ErrorCode' on error
+    /// - `key`: The key buffer
+    /// - `ret_buf`: The ret_buf buffer
+    fn get_value_complete(
+        &self,
+        result: Result<(), ErrorCode>,
+        key: &'static mut K,
+        ret_buf: SubSliceMut<'static, u8>,
+    );
+
+    /// This callback is called when the invalidate_key operation completes.
+    ///
+    /// - `result`: Nothing on success, 'ErrorCode' on error
+    /// - `key`: The key buffer
+    fn invalidate_key_complete(&self, result: Result<(), ErrorCode>, key: &'static mut K);
+
+    /// This callback is called when the garbage_collect operation completes.
+    ///
+    /// - `result`: Nothing on success, 'ErrorCode' on error
+    fn garbage_collect_complete(&self, result: Result<(), ErrorCode>);
+}
+
+pub trait KVSystem<'a> {
+    /// The type of the hashed key. For example `[u8; 8]`.
+    type K: KeyType;
+
+    /// Set the client.
+    fn set_client(&self, client: &'a dyn KVSystemClient<Self::K>);
+
+    /// Generate key.
+    ///
+    /// - `unhashed_key`: A unhashed key that should be hashed.
+    /// - `key_buf`: A buffer to store the hashed key output.
+    ///
+    /// On success returns nothing.
+    /// On error the unhashed_key, key_buf and `Result<(), ErrorCode>` will be returned.
+    fn generate_key(
```

Comment:
I think this is a broad general design question about a KV store interface and implementation, so maybe it shouldn't be part of this PR, _but_...

I don't understand as either an implementer or user of this interface what `generate_key` is supposed to do. Is the resulting key guaranteed to be unique for a given input key? Do I always need to use it? Is it allowed to be the identity function?

Also, why not return the resulting key instead of passing in a mutable "buffer," c-style? In particular, if it's being hashed I expect `Self::K` to _always_ be fixed size (either a usize, or a fixed-length array or something)

Reply:
> I think this is a broad general design question about a KV store interface and implementation, so maybe it shouldn't be part of this PR, _but_...

I think KV interface questions are within scope. The PRs that have already been merged got things to a working state.

> I don't understand as either an implementer or user of this interface what `generate_key` is supposed to do. Is the resulting key guaranteed to be unique for a given input key? Do I always need to use it? Is it allowed to be the identity function?

Currently no guarantees of uniqueness. You don't have to use it in the sense that if you already have a `K` type object than you can use that, but somehow you need to get `K`. It can be any function.

> Also, why not return the resulting key instead of passing in a mutable "buffer," c-style? In particular, if it's being hashed I expect `Self::K` to _always_ be fixed size (either a usize, or a fixed-length array or something).

I think the expectation is that it could be passed to a hardware peripheral.